rfd/0003-workspace-ram-management.md

```diff
@@ -0,0 +1,70 @@
+
+### Manual ballooning
+
+Previously, the only option for dynamic VM RAM management was memory ballooning where the hypervisor might manually readjust the size of the VM's RAM. Unfortunately, this introduces a ton of complexity in the hypervisor and is a half measure, not a solution, as creating a RAM usage policy which actually works well is very hard. Even if a good policy was designed and implemented, the inflation/deflation of the balloon is slow, unreliable, and unsuited for our use case. We want VMs to access most of the Host resources like containers and quickly give them back when unneeded.
+
+### Free page reporting
+
+The solution to this problem is Free Page Reporting[^1], a feature in `virtio-balloon` available since the 5.7 kernel, which gives a primitive in which the VM might give back their unused RAM to the hypervisor without manual intervention. By default, free page reporting operates in 4MB segments on x86_64, limiting its effectiveness without other measures. Since the 5.14 kernel, it's possible to override the segment size and for free page reporting to operate on a per-page granularity[^2]. In my testing, free page reporting with `page_reporting.page_reporting_order=0` works well and has acceptable overhead.
+
+### Limiting page cache usage via `virtio-pmem` and DAX
+
+I tested this possibility, and this doesn't work well:
+
+- For a 0.5TB sparse disk, the RAM overhead for the VM is 8GB, which is prohibitively large.
+- No support for discard, which results in the sparse disk files always growing in size.
+- Weird performance characteristics and longer boot time[^3].
+
+### THP (Transparent Huge Pages)
+
+Will always be disabled by default in Hocus workspaces to increase the effectiveness of free page reporting. THP might be enabled on the host, but we recommend disabling it for production usage.
+
+### KSM (Kernel Same Page Merging)
+
+Workspaces running on a single machine will usually run the same kernel, IDE version, and edit the same code. This setup is ideal for KSM. We recommend enabling KSM on the host machine, but it's not a requirement for Hocus.
+
+### DAMON (Data Access MONitor)
+
+For free page reporting to be effective, the workspace needs the ability to drop page caches it doesn't need. By default, Linux is designed to consume as much RAM as possible and never release it back until there is some memory pressure. Hocus workspaces use DAMON RECLAIM, available since the 5.16 kernel[^4], to proactively reclaim unused page caches. In our testing, this releases most of the caches the VM uses, but as DAMON is a very new subsystem in the kernel, there is still a lot of room for improvements[^5].
```

Comment:
```suggestion
For free page reporting to be effective, the workspace needs the ability to drop page caches it doesn't need. By default, Linux is designed to consume as much RAM as possible and never release it back until there is some memory pressure. Hocus workspaces use DAMON RECLAIM, available since the 5.16 kernel[^4], to proactively reclaim unused page caches. In our testing, this releases most of the caches the VM uses, but as DAMON is a very new subsystem in the kernel, there is still a lot of room for improvement[^5].
```